src/services/utilities.ts

```diff
@@ -1158,6 +1158,35 @@ namespace ts {
         }
     }
 
+    /**
+     * Psuedo-binary searches the list of children for the first element whose end position exceeds the input position
+     * Returns `undefined` if no element satisfies the condition, or the index of the satisfying element otherwise.
+     * This could probably be written using our `binarySeach` helper, but the `keyComparator` would be hairy enough
+     * that it's probably worth just writing out the algorithm in full.
+     * @param {number} start - Start search index, inclusive
+     * @param {number} end - End search index, inclusive
+     */
+    function searchChildrenSlice(position: number, children: Node[], start = 0, end = children.length - 1): number | undefined {
+        const pivot = Math.floor((end - start)/2) + start;
+        if (!children[pivot]) {
+            return undefined;
+        }
+        if (position < children[pivot].end) {
+            // first element whose end position is greater than the input position
+            if (!children[pivot - 1] || position >= children[pivot - 1].end) {
+                return pivot;
+            }
+            if (pivot === start) {
+                return undefined;
+            }
+            return searchChildrenSlice(position, children, start, pivot - 1);

```

Comment:
Seems like it would be ironic to blow the stack doing this and unrolling the recursion looks trivial in this case.

Reply:
Well, log_2 of 20000 is about 15, so I'd say it's unlikely; but I'll unroll it all the same.